File: micropython/payloads/device_info.py
"""设备信息 payload 编解码。"""

import struct

DEVICE_INFO_FIXED_SIZE_V1 = 29 # 设备信息 V1 固定头 payload 长度，不含变长字符串
DEVICE_INFO_FIXED_SIZE = DEVICE_INFO_FIXED_SIZE_V1 # 当前设备信息固定头 payload 长度


def _encode_var_string(value):
    """编码变长 UTF-8 字符串，单段最长 255 字节。"""
    return (value or '').encode('utf-8')[:0xFF]


def _decode_var_string(data):
    """解码变长 UTF-8 字符串。"""
    return data.decode('utf-8', 'replace')


class device_info_t:
    """设备信息 payload：固定头 + 变长字符串。"""

    FIXED_SIZE = DEVICE_INFO_FIXED_SIZE
    FORMAT = '<HHIHHHHBBBIBBBBBB'

    @staticmethod
    def pack(fields):
        """打包设备信息 payload。

        注意：manufacturer/model/serial 均按 255 字节截断；serial 不应为空。
        """
        manufacturer = _encode_var_string(fields.get('manufacturer', ''))
        model = _encode_var_string(fields.get('model', ''))
        serial = _encode_var_string(fields.get('serial', ''))
        values = ()
        if DEVICE_INFO_FIXED_SIZE >= DEVICE_INFO_FIXED_SIZE_V1:
            values = (
                fields['company_id'],
                fields['product_id'],
                fields['feature_mask'],
                fields['protocol_version'],
                fields['protocol_min_version'],
                fields['ota_max_kb'],
                fields['ota_chunk_max'],
                fields['firmware_major'],
                fields['firmware_minor'],
                fields['firmware_patch'],
                fields['firmware_build_id'],
                fields['hardware_major'],
                fields['hardware_minor'],
                fields['hardware_patch'],
                len(manufacturer),
                len(model),
                len(serial),
            )
        fixed = struct.pack(
            device_info_t.FORMAT,
            *values,
        )
        return fixed + manufacturer + model + serial

    @staticmethod
    def unpack(data):
        """解包设备信息 payload。

        注意：按当前 SDK 已知长度读取前缀，不根据协议版本切换解析结构。
        """
        if len(data) < DEVICE_INFO_FIXED_SIZE_V1:
            return None
        values = struct.unpack(device_info_t.FORMAT, data[:DEVICE_INFO_FIXED_SIZE_V1])
        manufacturer_len = values[14]
        model_len = values[15]
        serial_len = values[16]
        total_len = DEVICE_INFO_FIXED_SIZE_V1 + manufacturer_len + model_len + serial_len
        if len(data) < total_len:
            return None
        offset = DEVICE_INFO_FIXED_SIZE_V1
        manufacturer = _decode_var_string(data[offset: offset + manufacturer_len])
        offset += manufacturer_len
        model = _decode_var_string(data[offset: offset + model_len])
        offset += model_len
        serial = _decode_var_string(data[offset: offset + serial_len])
        result = {
            'company_id': values[0],
            'product_id': values[1],
            'feature_mask': values[2],
            'protocol_version': values[3],
            'protocol_min_version': values[4],
            'ota_max_kb': values[5],
            'ota_chunk_max': values[6],
            'firmware_major': values[7],
            'firmware_minor': values[8],
            'firmware_patch': values[9],
            'firmware_build_id': values[10],
            'hardware_major': values[11],
            'hardware_minor': values[12],
            'hardware_patch': values[13],
            'manufacturer': manufacturer,
            'model': model,
            'serial': serial,
        }
        return result

File: c/payloads/device_info.c
#include "device_info.h"

#include <string.h>

static void write_u16_le(uint8_t *buf, uint16_t offset, uint16_t value) {
    buf[offset] = (uint8_t)(value & 0xFF);
    buf[offset + 1] = (uint8_t)(value >> 8);
}

static uint16_t read_u16_le(const uint8_t *buf, uint16_t offset) {
    return (uint16_t)buf[offset] | ((uint16_t)buf[offset + 1] << 8);
}

static void write_u32_le(uint8_t *buf, uint16_t offset, uint32_t value) {
    buf[offset] = (uint8_t)(value & 0xFF);
    buf[offset + 1] = (uint8_t)((value >> 8) & 0xFF);
    buf[offset + 2] = (uint8_t)((value >> 16) & 0xFF);
    buf[offset + 3] = (uint8_t)((value >> 24) & 0xFF);
}

static uint32_t read_u32_le(const uint8_t *buf, uint16_t offset) {
    return (uint32_t)buf[offset] |
           ((uint32_t)buf[offset + 1] << 8) |
           ((uint32_t)buf[offset + 2] << 16) |
           ((uint32_t)buf[offset + 3] << 24);
}

static uint8_t bounded_len(uint8_t len) {
    return len;
}

bool device_info_pack(const device_info_t *fields, uint8_t *buf,
                      uint16_t buf_size, uint16_t *out_len) {
    if (!fields || !buf || !out_len) return false;
    uint8_t manufacturer_len = bounded_len(fields->manufacturer_len);
    uint8_t model_len = bounded_len(fields->model_len);
    uint8_t serial_len = bounded_len(fields->serial_len);
    if ((manufacturer_len > 0 && !fields->manufacturer) ||
        (model_len > 0 && !fields->model) ||
        (serial_len > 0 && !fields->serial)) {
        return false;
    }
    uint16_t needed = (uint16_t)(DEVICE_INFO_FIXED_SIZE + manufacturer_len + model_len + serial_len);
    if (buf_size < needed) return false;

    if (DEVICE_INFO_FIXED_SIZE >= DEVICE_INFO_FIXED_SIZE_V1) {
        write_u16_le(buf, 0, fields->company_id);
        write_u16_le(buf, 2, fields->product_id);
        write_u32_le(buf, 4, fields->feature_mask);
        write_u16_le(buf, 8, fields->protocol_version);
        write_u16_le(buf, 10, fields->protocol_min_version);
        write_u16_le(buf, 12, fields->ota_max_kb);
        write_u16_le(buf, 14, fields->ota_chunk_max);
        buf[16] = fields->firmware_major;
        buf[17] = fields->firmware_minor;
        buf[18] = fields->firmware_patch;
        write_u32_le(buf, 19, fields->firmware_build_id);
        buf[23] = fields->hardware_major;
        buf[24] = fields->hardware_minor;
        buf[25] = fields->hardware_patch;
        buf[26] = manufacturer_len;
        buf[27] = model_len;
        buf[28] = serial_len;
    }

    uint16_t offset = DEVICE_INFO_FIXED_SIZE;
    if (manufacturer_len > 0) {
        memcpy(buf + offset, fields->manufacturer, manufacturer_len);
        offset += manufacturer_len;
    }
    if (model_len > 0) {
        memcpy(buf + offset, fields->model, model_len);
        offset += model_len;
    }
    if (serial_len > 0) {
        memcpy(buf + offset, fields->serial, serial_len);
    }
    *out_len = needed;
    return true;
}

bool device_info_unpack(const uint8_t *data, uint16_t len,
                        device_info_t *out_fields) {
    if (!data || !out_fields || len < DEVICE_INFO_FIXED_SIZE_V1) return false;
    uint8_t manufacturer_len = data[26];
    uint8_t model_len = data[27];
    uint8_t serial_len = data[28];
    uint16_t needed = (uint16_t)(DEVICE_INFO_FIXED_SIZE_V1 + manufacturer_len + model_len + serial_len);
    if (len < needed) return false;

    if (len >= DEVICE_INFO_FIXED_SIZE_V1) {
        out_fields->company_id = read_u16_le(data, 0);
        out_fields->product_id = read_u16_le(data, 2);
        out_fields->feature_mask = read_u32_le(data, 4);
        out_fields->protocol_version = read_u16_le(data, 8);
        out_fields->protocol_min_version = read_u16_le(data, 10);
        out_fields->ota_max_kb = read_u16_le(data, 12);
        out_fields->ota_chunk_max = read_u16_le(data, 14);
        out_fields->firmware_major = data[16];
        out_fields->firmware_minor = data[17];
        out_fields->firmware_patch = data[18];
        out_fields->firmware_build_id = read_u32_le(data, 19);
        out_fields->hardware_major = data[23];
        out_fields->hardware_minor = data[24];
        out_fields->hardware_patch = data[25];
        out_fields->manufacturer_len = manufacturer_len;
        out_fields->model_len = model_len;
        out_fields->serial_len = serial_len;
    }

    uint16_t offset = DEVICE_INFO_FIXED_SIZE_V1;
    out_fields->manufacturer = (const char *)(data + offset);
    offset += manufacturer_len;
    out_fields->model = (const char *)(data + offset);
    offset += model_len;
    out_fields->serial = (const char *)(data + offset);
    return true;
}
